ibis/file/hdf/client.py

```diff
@@ -0,0 +1,73 @@
+import pandas as pd
+import ibis.expr.operations as ops
+from ibis.file.client import Client
+from ibis.file.utils import pathlib
+from ibis.pandas.core import data_preload, execute
+from ibis.pandas.client import pandas_dtypes_to_ibis_schema
+
+
+class HDFTable(ops.DatabaseTable):
+    pass
+
+
+class HDFClient(Client):
+    extension = 'h5'
+
+    def write_to_file(self, path, key, t, format='table',
+                      data_columns=True, **kwargs):
+
+        path = self.root / path
+        data = execute(t)
+        data.to_hdf(str(path), key, format=format,
+                    data_columns=data_columns, **kwargs)
+
+    def table(self, name, path):
+        if name not in self.list_tables(path):
+            raise AttributeError(name)
+
+        # get the schema
+        with pd.HDFStore(str(path), mode='r') as store:
+            df = store.select(name, start=0, stop=0)
+            schema = pandas_dtypes_to_ibis_schema(df, {})
+
+        t = HDFTable(name, schema, self).to_expr()
+        self.dictionary[name] = path
+        return t
+
+    def list_tables(self, path=None):
+        # tables are individual tables within a file
+
+        if path is None:
+            path = self.root
+
+        if (path.is_file() and str(path).endswith(self.extension)):
+
+            with pd.HDFStore(str(path), mode='r') as store:
+                # strip leading /
+                return [k[1:] for k in store.keys()]
+
+        return []
+
+    def list_databases(self, path=None):
```

Comment:
This looks very similar to the csv version. Is there a way to factor this out into `FileClient`?

Reply:
its actually similar to the list_tables in csv. But databases in hdf5 are actually files, the tables are inside, so we have to traverse a bit differntly.